### PR TITLE
MAISTRA-1076 add upgrade/downgrade validation

### PR DESCRIPTION
**Pull request by rcernich**
Signed-off-by: rcernich <rcernich@redhat.com>

### CODE REVIEW

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (rcernich) [inline pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: There are a couple of places where we do things like this.  Should we?

My concerns are with the number of calls to the apiserver and possibly the amount of data returned.  We query a number of deprecated resource types to ensure an upgrade won't break (we could remove that logic), and we query a number of resources to see if they're configured in an unsupported manner (e.g. http- named ports using 443).

File: pkg/controller/servicemesh/webhooks/mutation/controlplane.go
Comment (luksa) [inline pkg/controller/servicemesh/webhooks/mutation/controlplane.go]: Do we really need to set the version on update? This code won't be invoked until the user updates the SMCP, which means that the rest of the code always needs to treat an empty version as legacy. So if we need to do this everywhere else, what is the point of having this code here?

File: pkg/controller/servicemesh/webhooks/validation/controlplane.go
Comment (luksa) [inline pkg/controller/servicemesh/webhooks/validation/controlplane.go]: These three lines can be removed, since we already check if the (new) Object is deleted. If the new Object is not deleted, then the old Object also can't be.

File: pkg/controller/servicemesh/webhooks/validation/controlplane.go
Comment (luksa) [inline pkg/controller/servicemesh/webhooks/validation/controlplane.go]: nit: unnecessary blank line

File: pkg/controller/servicemesh/webhooks/validation/controlplane.go
Comment (luksa) [inline pkg/controller/servicemesh/webhooks/validation/controlplane.go:184]: Context needs to be passed in from calling function.

File: pkg/controller/servicemesh/webhooks/validation/controlplane_test.go
Comment (luksa) [inline pkg/controller/servicemesh/webhooks/validation/controlplane_test.go]: Oooh, I like this approach. How about reducing the boilerplate even further by introducing a function like this:
```go
func setNestedField(obj map[string]interface{}, path string, value interface{}) {
    unstructured.SetNestedField(obj, value, strings.Split(path, ".")...)
}
```

This would make the tests even easier to read:
```go
  setNestedField(smcp.Spec.Istio, "global.proxy.alwaysInjectSelector", false)
```

File: pkg/controller/common/helm.go
Comment (dgn) [inline pkg/controller/common/helm.go]: nit: I know this is old, but I think this is no longer hard-coded - so the FIXME can be removed :)

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (dgn) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: Should we maybe use `istio/client-go` to query the objects instead of unstructured?

File: pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go
Comment (dgn) [inline pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go:98]: Is this a meaningful check? This is more of a general configuration issue than something that comes up during upgrade, right?

File: pkg/controller/servicemesh/webhooks/validation/controlplane_v1_0.go
Comment (dgn) [inline pkg/controller/servicemesh/webhooks/validation/controlplane_v1_0.go]: I think this is a good start for better validation!

File: pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go
Comment (luksa) [inline pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go]: I'm not sure we should be rejecting the upgrade if these resources are present. By doing so, we're forcing users to first delete these resources and then update the SMCP.  But users usually update multiple resources in a single kubectl apply command. Plus, doesn't the operator itself create some of these resources when you create the SMCP? Does this mean they have to delete them manually even though they never created them?

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (luksa) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: We should use caching. It's not appropriate for a webhook to produce multiple API requests. Also consider one of my other comments: should we really check for the existence of other resource types?

File: pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go:98]: I guess it depends on your general philosophy of what we should be checking.  If the user is using http/http2 protocols on port 443, an upgrade will break their application.  The logic in these upgrade/downgrade checks is intended to ensure an upgrade or downgrade won't break their applications.  I think it would be beneficial for the operator to migrate things like this when possible (nothing in 1.0 <-> 1.1, but I could see migrating Policy to AuthorizationPolicy in the future).

File: pkg/controller/servicemesh/webhooks/mutation/controlplane.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/mutation/controlplane.go]: What if the user removes the `Version` field during an update?  I think this will need to be reset to the previously `AppliedVersion` or else a `1.0` control plane will be installed, whether they want that or not.  Regarding the legacy initialization, I think it does help, as currently, reconciliation only occurs when the SMCP resource has been modified, so this would simplify reconciliation.  However, if we add code to trigger reconciliation in response to user modifications of managed resources, we would need to ensure empty values for `Version` are supported in the reconciler.

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: We should, but I was weary of adding that as a dependency.  I'll see how much effort is involved.

File: pkg/controller/servicemesh/webhooks/validation/controlplane.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane.go]: I'll remove this.

File: pkg/controller/servicemesh/webhooks/validation/controlplane.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane.go:184]: I'll update this.

File: pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go]: Good question.  First, controller managed resources are filtered from the check, so those will not prevent an upgrade.  Regarding user created CRs, I did not check to see if there was an upgrade path.  These were deprecated in Istio 1.2 and removed in Istio 1.3, and since we're moving from Istio 1.1 to 1.4, it's likely there isn't anything in Istio 1.1 that supports migration to a new resource.  Is it possible to return a warning from validation?

File: pkg/controller/servicemesh/webhooks/mutation/controlplane.go
Comment (luksa) [reply pkg/controller/servicemesh/webhooks/mutation/controlplane.go]: Oh, I wouldn't allow unsetting the version, once it's set. I mean, the mutating webhook should default the version to the one in the oldObject (if set).

Also, I do think we'll start triggering reconciliation much more frequently than we do now. It's really not ok to fix the owned resources only when the user updates the SMCP. That goes against the basic idea of declarative configuration.

File: pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go:98]: I just realized that we do check for `AuthorizationPolicy` on downgrade.  Should we add logic to migrate that to Policy/RbacConfig on downgrade?

File: pkg/controller/servicemesh/webhooks/mutation/controlplane.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/mutation/controlplane.go]: Okay, I'll modify this to initialize it to the `oldSmcp.Spec.Version` and leave it at that.  This will mean that `Version == ""` will imply `v1.0`.

Comment (luksa) [reply pkg/controller/servicemesh/webhooks/mutation/controlplane.go]: We could add a `GetEffectiveVersion()` to the SMCP struct to centralize this, but I don't know how we'd force ourselves to use the function instead of simply reading the field.

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: These types have been removed from istio/client-go@1.4.5, so we're stuck using unstructured.

**Comment by rcernich** [comment]
Ready for another review.

I was unable to integrate istio/client-go, as the deprecated resources are not available in 1.4.5.
Version is now defaulted to the old version if it's blank in the updated resource.

**Comment by luksa** [comment]
> Version is now defaulted to the old version if it's blank in the updated resource.

Do we still need the `AppliedVersion` field now?

**Comment by rcernich** [comment]
re `AppliedVersion`...good question.  I think so, as this is the only way the reconciler knows whether or not it is performing an upgrade of the control plane.  That said, we don't actual have any special upgrade/downgrade logic, so it wouldn't be used.

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (dgn) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: I see, thanks for checking!

**Comment by rcernich** [comment]
@luksa, @dgn, @knrc, did we want to merge this for 1.1?

**Comment by dgn** [comment]
> @luksa, @dgn, @knrc, did we want to merge this for 1.1?

I'll test this for a bit today

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (dgn) [inline pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: The operator doesn't have permissions to list these

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (dgn) [inline pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: should be `in v1.0`

File: pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go
Comment (dgn) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go:98]: > I guess it depends on your general philosophy of what we should be checking. If the user is using http/http2 protocols on port 443, an upgrade will break their application. 

Didn't know this was the case! Makes total sense to check this then, sorry

**Review by dgn — COMMENTED** [review]
Found one permission issue.

**Review by dgn — APPROVED** [review]
I think this is good to go as soon as the permission issue is resolved, I also consider it low-risk enough to be merged at this time. Will let final vote go to @knrc and @luksa. Thanks Rob!

File: pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go
Comment (dgn) [inline pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go]: this doesn't catch the case where the port is simply named `http` or `http2` (which is what I used for testing, coincidentally)

Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_upgrade_v1_0.go]: Thanks @dgn.  I'll update this and add a test case for it.

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: will fix

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: will update the roles

Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: Side note, should we look at deprecating ServiceMeshRbacConfig (our version of ClusterRbacConfig, which was replaced by AuthorizationPolicy) in favor of using AuthorizationPolicy?

**Comment by rcernich** [comment]
I've processed @dgn's last round of comments and am currently looking at manually adding typed resources for the istio types used in the upgrade/downgrade validation, to get cache support from the controller-runtime client.

File: pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go
Comment (knrc) [reply pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go]: +1, rbac should be deprecated in 1.4

**Comment by rcernich** [comment]
This last round of changes integrates @dgn's work for MAISTRA-1160 into 1.0 validation.  Validation will only be performed on 1.0 SMCP resources when dowgrading from 1.1.  I think this should be safe, because an upgrade would be prevented if those settings are misconfigured.  The only setting that would prevent downgrading is the use of `kiali.jaegerInClusterURL` and some other 1.1 specific settings.  Users can remove the offending settings when downgrading.

(This topic came up because of MAISTRA-1160.  In 1.0 it might have been possible for users to use various configurations with zipkin.address set, which are flagged as invalid in 1.1.  Validation for these cases is only checked in 1.0 when downgrading from 1.1, and since they can't upgrade with a misconfigured system, downgrading shouldn't be a problem either.)

**Comment by rcernich** [comment]
This also includes some minor tweaks to the default logic, as we were treating `smcp.Spec.Version == ""` as `v1.1`, when it should be treated as legacy `v1.0`.  In other words, whenever `Version` is not set, it should be assumed to be `v1.0`, always.  This was not consistent before.  (Note, the mutating webhook will initialize `Version` to the latest version, which is `v1.1` today, i.e. defaulting occurs in the mutating webhook.)

**Comment by rcernich** [comment]
I've verified that the "typed" resources used in the validation check work in a real install, so this should be good to go. @knrc, @luksa, it requires your final blessing before it can be merged.  Please review and approve if you are satisfied.

**Review by knrc — APPROVED** [review]

**Pull request merged**
